File: src/binpl.jl
using Printf

# Constants
const BINPL_VERSION = "v1.0-preview1"

# Function prototypes
function interpret_binary(binary_code::String, decimal_mode::Bool, hexadecimal_mode::Bool)
    len = length(binary_code)

    # Make sure the length is a multiple of 8
    if len % 8 != 0
        println("Error: Invalid binary string length")
        exit(1)
    end

    # Iterate through the binary string, interpreting each group of 8 bits
    for i in 1:8:len
        decimal_value = 0
        for j in 0:7
            decimal_value = (decimal_value << 1) | (parse(Int, binary_code[i + j]) - 0)
        end

        if decimal_mode
            print(decimal_value, " ")
        elseif hexadecimal_mode
            @printf("%02X ", decimal_value)
        else
            print(Char(decimal_value))
        end
    end
    println()
end

# Main function
function main(args)
    if length(args) != 1 && length(args) != 2
        println("Usage: binpl [binary file] [-d or --decimal] [-h or --hexadecimal]")
        return
    end

    # Check for help option
    if length(args) == 1 && (args[1] == "-h" || args[1] == "--help")
        println("Usage: binpl [binary file] [-d or --decimal] [-h or --hexadecimal]")
        println("       -v or --version       Version information.")
        println("       -h or --help          Usage information.")
        println("       -d or --decimal       Interpret binary as decimal numbers.")
        println("       -h or --hexadecimal   Interpret binary as hexadecimal numbers.")
        println("       --author              Author information.")
        return
    end

    if length(args) == 1 && (args[1] == "-v" || args[1] == "--version")
        println(BINPL_VERSION)
        return
    end

    if length(args) == 1 && (args[1] == "--author")
        println("Copyright (c) 2024 Cyril John Magayaga")
        return
    end

    # Open the binary file
    binary_code = read(args[1], String)
    decimal_mode = false # Default mode: ASCII interpretation
    hexadecimal_mode = false # Default mode: ASCII interpretation

    if length(args) == 2
        if args[2] == "-d" || args[2] == "--decimal"
            decimal_mode = true
        elseif args[2] == "-h" || args[2] == "--hexadecimal"
            hexadecimal_mode = true
        else
            println("Error: Invalid option '", args[2], "'")
            return
        end
    end

    # Parse binary code, skipping comments
    for line in split(binary_code, '\n')
        if !startswith(line, ";;")
            interpret_binary(String(line), decimal_mode, hexadecimal_mode)  # Convert SubString{String} to String
        end
    end
end

# Run the main function with command-line arguments
main(ARGS)

File: src/binpl.lua
--
-- BINPL (v1.0-preview1 / July 12, 2024)
-- Copyright (c) 2024 Cyril John Magayaga
--

BINPL_VERSION = "v1.0-preview1"

-- Function to interpret binary code
function interpretBinary(binaryCode, decimalMode, hexadecimalMode)
    local len = #binaryCode

    -- Make sure the length is a multiple of 8
    if len % 8 ~= 0 then
        print("Error: Invalid binary string length")
        os.exit(1)
    end

    -- Iterate through the binary string, interpreting each group of 8 bits
    for i = 1, len, 8 do
        local decimalValue = 0
        for j = 0, 7 do
            decimalValue = (decimalValue << 1) | (binaryCode:sub(i + j, i + j) - '0')
        end
        
        if decimalMode then
            io.write(decimalValue .. " ")
        
        elseif hexadecimalMode then
            io.write(string.format("%02X ", decimalValue))
        
        else
            io.write(string.char(decimalValue))
        end
    end
    io.write("\n")
end

-- Main function
function main(args)
    if #args < 1 or #args > 2 then
        print("Usage: binpl [binary file] [-d or --decimal] [-h or --hexadecimal]")
        return
    end

    -- Check for help option
    if #args == 1 and (args[1] == "-h" or args[1] == "--help") then
        print("Usage: binpl [binary file] [-d or --decimal] [-h or --hexadecimal]")
        print("       -v or --version       Version information.")
        print("       -h or --help          Usage information.")
        print("       -d or --decimal       Interpret binary as decimal numbers.")
        print("       -h or --hexadecimal   Interpret binary as hexadecimal numbers.")
        print("       --author              Author information.")
        return
    end

    if #args == 1 and (args[1] == "-v" or args[1] == "--version") then
        print(BINPL_VERSION)
        return
    end

    if #args == 1 and (args[1] == "--author") then
        print("Copyright (c) 2024 Cyril John Magayaga")
        return
    end

    local file = io.open(args[1], "rb")
    if not file then
        print("Error: Unable to open file '" .. args[1] .. "'")
        return
    end

    local binaryCode = file:read("*a")
    file:close()

    local decimalMode = false -- Default mode: ASCII interpretation
    local hexadecimalMode = false -- Default mode: ASCII interpretation

    if #args == 2 then
        if args[2] == "-d" or args[2] == "--decimal" then
            decimalMode = true
        
        elseif args[2] == "-h" or args[2] == "--hexadecimal" then
            hexadecimalMode = true
        
        else
            print("Error: Invalid option '" .. args[2] .. "'")
            return
        end
    end

    -- Parse binary code, skipping comments
    for line in binaryCode:gmatch("[^\r\n]+") do
        if line:sub(1, 2) ~= ";;" then
            interpretBinary(line, decimalMode, hexadecimalMode)
        end
    end
end

main(arg)